File: Backtracking/combination_sum_LC39.py
class Solution():
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []

        def bckrtrck(candidates, target, arr):
            if target == 0:
                res.append(arr)
            for i in range(len(candidates)):
                if candidates[i] > target:
                    break
                bckrtrck(candidates[i:], target -
                         candidates[i], arr + [candidates[i]])
        bckrtrck(sorted(candidates), target, [])
        return res


if __name__ == '__main__':
    obj = Solution()
    candidates = [[2, 3, 6, 7], [2, 3, 5]]
    target = [7, 8]
    ans = [
        [[2, 2, 3], [7]],
        [[2, 2, 2, 2], [2, 3, 3], [3, 5]]
    ]

    for i in range(len(candidates)):
        assert(obj.combinationSum(candidates[i], target[i]) == ans[i])

File: Backtracking/CombinationSumLC39.java
import java.util.*;

class SolutionLC39 {
    public List<List<Integer>> combinationSum(final int[] candidates, final int target) {
        Arrays.sort(candidates);

        final List<List<Integer>> result = new ArrayList<List<Integer>>();

        helper(candidates, target, new ArrayList<Integer>(), result, 0);

        return result;
    }

    public void helper(final int[] candidates, final int target, final ArrayList<Integer> arr,
            final List<List<Integer>> result, final int start) {

        if (target == 0) {
            result.add(new ArrayList<Integer>(arr));
        } else {
            for (int index = start; index < candidates.length; index++) {
                if (candidates[index] > target)
                    return;
                arr.add(candidates[index]);
                helper(candidates, target - candidates[index], arr, result, index);
                arr.remove(arr.size() - 1);
            }
        }
    }
}

public class CombinationSumLC39 {

    public static void main(final String[] args) {
        SolutionLC39 solution = new SolutionLC39();

        int candidates[] = { 2, 3, 5 };
        int target = 8;

        System.out.println(solution.combinationSum(candidates, target));
    }

}